### PR TITLE
XmlDocument: save BOM-less UTF-8 files by default

### LINKED ISSUE #28218: System.Xml.XmlDocument.Save() unexpectedly creates UTF-8 files *with BOM*
`System.Xml.XmlDocument.Save()`, when given a file path to write to, unexpectedly creates UTF-8 encoded files _with BOM_ if the document's XML declaration has an `encoding="UTF-8"` attribute.

(By contrast, the absence of the `encoding` attribute or absence of an XML declaration altogether causes a BOM-_less_ UTF-8 file to be created, as expected.)

Repro in PowerShell Core (on any supported platform; PowerShell's `[xml]` cast instantiates a `System.Xml.XmlDocument` from the specified XML string):

```powershell
([xml] '<?xml version="1.0" encoding="UTF-8" ?><foo />').Save("$HOME/t.xml")
Get-Content -AsByteStream $HOME/t.xml | Select-Object -First 3
```

The above yields:

```none
239
187
191
```

which are the byte values of the UTF-8 BOM (`0xef 0xbb 0xbf`)

### PR DESCRIPTION
**Pull request by mklement0**
Fix #28218

### CODE REVIEW

File: src/libraries/System.Private.Xml/tests/XmlDocument/XmlDocumentTests/SaveTests.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO;
+using System.Text;
+using Xunit;
+
+namespace System.Xml.Tests
+{
+  public class SaveTests
```
Comment (krwq) [inline src/libraries/System.Private.Xml/tests/XmlDocument/XmlDocumentTests/SaveTests.cs:10]: please match the indentations in this file to other files in the project

File: src/libraries/System.Private.Xml/tests/XmlDocument/XmlDocumentTests/SaveTests.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  {
+    public static readonly byte[] utf8BomBytes = new byte[] { 0xef, 0xbb, 0xbf };
```
Comment (krwq) [inline src/libraries/System.Private.Xml/tests/XmlDocument/XmlDocumentTests/SaveTests.cs:12]: you can get these with `Encoding.UTF8.GetPreamble()`

File: src/libraries/System.Private.Xml/tests/XmlDocument/XmlDocumentTests/SaveTests.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      xmlDocument.AppendChild(xmlDocument.CreateElement("Foo"));
+
+      XmlWriterSettings settings = new XmlWriterSettings();
+      // Set the encoding explicitly, to UTF-8 with BOM.
+      settings.Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true);
+
+      var tmpFile = Path.GetTempFileName();
+      try
+      {
+        using (XmlWriter writer = XmlWriter.Create(tmpFile, settings))
```
Comment (krwq) [inline src/libraries/System.Private.Xml/tests/XmlDocument/XmlDocumentTests/SaveTests.cs:50]: write to a memory stream instead

File: src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs [hunk 4]
```diff
@@ -1378,7 +1378,9 @@ public virtual void LoadXml(string xml)
                     string value = Declaration.Encoding;
                     if (value.Length > 0)
                     {
-                        return System.Text.Encoding.GetEncoding(value);
+                        return string.Equals(value, "utf-8", StringComparison.OrdinalIgnoreCase) ?
+                                    new UTF8Encoding(encoderShouldEmitUTF8Identifier: false) :
```
[the remark this answers is not in the document]
Comment (mklement0) [reply src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs:1382]: @Wraith2, given that `.GetEncoding(value)` was previously called _unconditionally_, I don't see a need for this change - but I'll defer to @krwq as to whether an optimization is called for in this case.

[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs:1382]: If there is something handy you might use it but in general creating this type will be relatively cheap comparing to loading XML

**Comment by jeffhandley** [comment]
@krwq It looks like your feedback has been addressed; can you take another look please?

**Comment by krwq** [comment]
@jeffhandley @mklement0 I'm waiting how https://github.com/dotnet/runtime/issues/51353 will play out first. If there is a consensus there:
- if they decide not to change it then we need to re-evaluate this PR with the same reasoning and see if we still want to change it
- if they decide to change it then this change won't be needed as you will get this behavior without any change

**Comment by krwq** [comment]
at the same time PR looks good to me

**Comment by jeffhandley** [comment]
I'm going to mark this as a draft for now since #51353 is still pending

**Pull request closed without merging**